src/pybind/mgr/dashboard/module.py

```diff
@@ -510,28 +514,32 @@ def serve(self):
         class Root(object):
             @cherrypy.expose
             def index(self):
-                active_uri = module.get_active_uri()
-                if active_uri:
-                    module.log.info("Redirecting to active '%s'", active_uri)
-                    raise cherrypy.HTTPRedirect(active_uri)
+                if module.get_localized_module_option('standby_behaviour', 'redirect') == 'redirect':
+                    active_uri = module.get_active_uri()
+                    if active_uri:
+                        module.log.info("Redirecting to active '%s'", active_uri)
+                        raise cherrypy.HTTPRedirect(active_uri)
+                    else:
+                        template = """
+                    <html>
+                        <!-- Note: this is only displayed when the standby
+                             does not know an active URI to redirect to, otherwise
+                             a simple redirect is returned instead -->
+                        <head>
+                            <title>Ceph</title>
+                            <meta http-equiv="refresh" content="{delay}">
+                        </head>
+                        <body>
+                            No active ceph-mgr instance is currently running
+                            the dashboard. A failover may be in progress.
+                            Retrying in {delay} seconds...
+                        </body>
+                    </html>
+                        """
+                        return template.format(delay=5)
                 else:
-                    template = """
-                <html>
-                    <!-- Note: this is only displayed when the standby
-                         does not know an active URI to redirect to, otherwise
-                         a simple redirect is returned instead -->
-                    <head>
-                        <title>Ceph</title>
-                        <meta http-equiv="refresh" content="{delay}">
-                    </head>
-                    <body>
-                        No active ceph-mgr instance is currently running
-                        the dashboard. A failover may be in progress.
-                        Retrying in {delay} seconds...
-                    </body>
-                </html>
-                    """
-                    return template.format(delay=5)
+                    status = module.get_localized_module_option('standby_error_status_code', 500)
```

Comment:
Mmm, this is the kind of thing I would't allow to be localized. Don't you think it'd cause a tricky behaviour for some nodes to redirect with 3xx while others fail with 5xx?

Reply:
Indeed, an admin is able to do weird things with this, but maybe it needs to be necessary in some cases. The ``get_localized_module_option`` falls back to the global setting if there is no localized one. The documentation never talks about the localized configuration capability, so we should be worried about weird usage. But IMO we should implement this feature using the localized API to be able to configure it the weird way if it's necessary.

Reply:
Mmm. Have to disagree here. Every new configuration option increases the degrees of freedom of a system, which involves a exponential increase in the end-to-end cases to be tested. The systems should be as flexible as strictly required and nothing more. In the end, this moves design/coding decision-making into customer's configuration management issues (a related paper from NetApp folks: ["Hey, you have given me too many knobs"](https://tianyin.github.io/papers/fse15.pdf)).

Compare test cases:
- 1 new option `standby_behaviour=["redirect"|"error"]`. 2 tests:
  1. `standby_behaviour="redirect"`
  1. `standby_behaviour "error"`

To: 
- 2 new localized options `standby_behaviour=["redirect"|"error"]` + `standby_error_status_code=[300,500]`' for 3-node mgr cluster. 3 ^ (1+2) = 27 tests:
  1. `standby_behaviour[mgr.x]="redirect"`, `standby_behaviour[mgr.y]="redirect"`, `standby_behaviour[mgr.z]="redirect"`
  1. `standby_behaviour[mgr.x]="redirect"`, `standby_behaviour[mgr.y]="redirect"`, `standby_behaviour[mgr.z]="error"` AND:
     1. `standby_error_status_code[mgr.z]=300`
     1. `standby_error_status_code[mgr.z]=500`
  1. `standby_behaviour[mgr.x]="redirect"`, `standby_behaviour[mgr.y]="error"`, `standby_behaviour[mgr.z]="error"` AND:
     1. `standby_error_status_code[mgr.z]=300`
        1. `standby_error_status_code[mgr.y]=300`
        1. `standby_error_status_code[mgr.y]=500`
     1. `standby_error_status_code[mgr.z]=500`
...

Reply:
I can not share your opinion because i've had many situations in my developer life where such 'too much knobs' saved the life of my company out in the field at the customer. But i will adapt the code to get this PR merged soon.

Reply:
That's great, @votdev, but can you please tell me how having different ceph-mgr instances returning for example: a 418 ("I'm a teapot"), 402 ("Payment required"), 505 ("HTTP version not supported") or 599 (??), all of them values that we accept here (and which we should never return even on error), are somehow going to help any customer?

I mean, we shouldn't do things "just in case". This PR has an issue ticket attached which clearly states the scope of changes: "support HAProxy with SSL-enabled dashboard". Can that be just handled with a single config option to choose between a 302 ("Found") or a 503 ("Service not available")? I think so. Then, IMHO everything else is added complexity with zero-value.


I mean, we may also allow for any HTTP code from 0 to 999. Some of them are undefined/wrong, but who knows?